File: proof/deleva/risk.py
"""Deterministic risk-state evaluation.

This module answers exactly one question -- "what is the current risk
state?" -- and does nothing else. It never reads chain state, never calls
Almanak or KeeperHub, and never executes anything. Phase 2's autonomous loop
calls this after reading a Position and before deciding whether to act.
"""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal, InvalidOperation

from deleva.models import Position, RiskState


class InvalidHealthFactorError(ValueError):
    """Raised when a health factor value is present but not a valid number."""


@dataclass(frozen=True)
class RiskEvaluator:
    """Evaluates a Position's health factor against a configured trigger threshold.

    Matches Aave's own convention: a health factor at or above the threshold
    is healthy; strictly below it is breached. The boundary is inclusive on
    the healthy side (health_factor == threshold -> HEALTHY), matching the
    spec this was built against.
    """

    trigger_threshold: Decimal

    def evaluate_health_factor(self, health_factor: Decimal | int | str | None) -> RiskState:
        if health_factor is None:
            return RiskState.UNKNOWN

        if isinstance(health_factor, Decimal):
            hf = health_factor
        else:
            try:
                hf = Decimal(health_factor)
            except (InvalidOperation, TypeError, ValueError) as exc:
                raise InvalidHealthFactorError(
                    f"health_factor={health_factor!r} is not a valid number. "
                    "If this came from a raw on-chain read, it must be normalized "
                    "with deleva.models.normalize_health_factor() first."
                ) from exc

        if hf < 0:
            raise InvalidHealthFactorError(f"health_factor={hf} cannot be negative")

        if hf < self.trigger_threshold:
            return RiskState.THRESHOLD_BREACHED
        return RiskState.HEALTHY

    def evaluate(self, position: Position) -> RiskState:
        return self.evaluate_health_factor(position.health_factor)

File: web/src/lib/risk.ts
// Presentation-only helper: given a health factor and the strategy's own
// trigger threshold (both already computed by the backend), decide how to
// label/color it. Does NOT reimplement DELEVA's risk evaluation -- it only
// mirrors the same HEALTHY / THRESHOLD_BREACHED boundary the backend's
// RiskEvaluator already applied, purely for display.

export type RiskTier = "healthy" | "danger" | "unknown";

export function riskTier(healthFactor: string | null, triggerThreshold: string | null): RiskTier {
  if (healthFactor === null || triggerThreshold === null) return "unknown";
  const hf = Number(healthFactor);
  const trigger = Number(triggerThreshold);
  if (Number.isNaN(hf) || Number.isNaN(trigger)) return "unknown";
  return hf >= trigger ? "healthy" : "danger";
}

export function riskLabel(tier: RiskTier): string {
  return tier === "healthy" ? "Healthy" : tier === "danger" ? "Threshold breached" : "Unavailable";
}
